File: servers/coordinates-mcp/src/bin/server/ownership.rs
use rmcp::{ErrorData as McpError, RoleServer, service::RequestContext};
use veoveo_coordinates_mcp::state::CoordinateScope;
use veoveo_mcp_contract::{GatewayInternalIdentity, PlaneCaller, PrincipalKind};
use veoveo_platform_store::PrincipalKind as StorePrincipalKind;
use veoveo_task_runtime::TaskOwner;

use super::app_state::AppState;

pub(super) fn internal_identity(
    context: &RequestContext<RoleServer>,
) -> Result<GatewayInternalIdentity, McpError> {
    let parts = context
        .extensions
        .get::<axum::http::request::Parts>()
        .ok_or_else(|| McpError::invalid_request("authenticated HTTP context missing", None))?;
    parts
        .extensions
        .get::<GatewayInternalIdentity>()
        .cloned()
        .ok_or_else(|| McpError::invalid_request("gateway identity missing", None))
}

pub(super) fn internal_caller(
    context: &RequestContext<RoleServer>,
) -> Result<PlaneCaller, McpError> {
    let parts = context
        .extensions
        .get::<axum::http::request::Parts>()
        .ok_or_else(|| McpError::invalid_request("authenticated HTTP context missing", None))?;
    let identity = parts
        .extensions
        .get::<GatewayInternalIdentity>()
        .cloned()
        .ok_or_else(|| McpError::invalid_request("gateway identity missing", None))?;
    let bearer = parts
        .extensions
        .get::<super::internal_auth::ForwardedBearer>()
        .map(|bearer| bearer.0.clone())
        .ok_or_else(|| McpError::invalid_request("forwarded bearer missing", None))?;
    Ok(caller_from(identity, bearer))
}

pub(super) fn caller_from(identity: GatewayInternalIdentity, bearer: String) -> PlaneCaller {
    let memberships = identity.principal.group_memberships();
    PlaneCaller {
        bearer_token: bearer,
        identity,
        memberships,
    }
}

pub(super) fn runtime_owner(identity: &GatewayInternalIdentity) -> TaskOwner {
    TaskOwner {
        principal_key: identity.principal.id.to_string(),
        principal_kind: match identity.principal.kind {
            PrincipalKind::User => veoveo_task_runtime::PrincipalKind::User,
            PrincipalKind::Service => veoveo_task_runtime::PrincipalKind::Service,
        },
        issuer: identity.principal.issuer.to_string(),
        subject: identity.principal.subject.to_string(),
        profile: identity.profile.to_string(),
        tenant_key: identity.principal.tenant.as_ref().map(ToString::to_string),
        data_labels: identity
            .principal
            .data_labels
            .iter()
            .map(ToString::to_string)
            .collect(),
    }
}

pub(super) async fn coordinate_scope_from_identity(
    state: &AppState,
    identity: &GatewayInternalIdentity,
) -> Result<CoordinateScope, McpError> {
    coordinate_scope_from_runtime(state, &runtime_owner(identity)).await
}

pub(super) async fn coordinate_scope_from_runtime(
    state: &AppState,
    owner: &TaskOwner,
) -> Result<CoordinateScope, McpError> {
    let identity = state
        .tasks
        .platform_store()
        .ensure_identity(
            owner.tenant_key(),
            &owner.principal_key,
            &owner.issuer,
            &owner.subject,
            match owner.principal_kind {
                veoveo_task_runtime::PrincipalKind::User => StorePrincipalKind::User,
                veoveo_task_runtime::PrincipalKind::Service => StorePrincipalKind::Service,
            },
        )
        .await
        .map_err(|error| McpError::internal_error(error.to_string(), None))?;
    Ok(CoordinateScope {
        identity,
        data_labels: owner.data_labels.clone(),
    })
}

pub(super) async fn optional_task_owner(
    state: &AppState,
    task_id: &str,
) -> Result<Option<TaskOwner>, McpError> {
    state
        .tasks
        .owner(task_id)
        .await
        .map_err(|error| McpError::internal_error(error.to_string(), None))
}

pub(super) async fn require_task_owner(
    state: &AppState,
    context: &RequestContext<RoleServer>,
    task_id: &str,
) -> Result<GatewayInternalIdentity, McpError> {
    let identity = internal_identity(context)?;
    let owner = optional_task_owner(state, task_id)
        .await?
        .ok_or_else(|| McpError::invalid_request("task ownership record missing", None))?;
    if task_owner_allows(&owner, &identity) {
        Ok(identity)
    } else {
        Err(McpError::invalid_request(
            "coordinates task policy denied request",
            None,
        ))
    }
}

pub(super) fn task_owner_allows(owner: &TaskOwner, identity: &GatewayInternalIdentity) -> bool {
    let caller = runtime_owner(identity);
    owner.allows(
        &caller.principal_key,
        &caller.profile,
        caller.tenant_key.as_deref(),
        &caller.data_labels,
    )
}

File: templates/python-mcp/src/datasheet_mcp/server/ownership.py
"""Principal, tenant, and label ownership for datasheet tasks."""

from __future__ import annotations

from typing import Any

from mcp.shared.exceptions import McpError
from mcp.types import INVALID_REQUEST, ErrorData

from veoveo_mcp.contract.identity import GatewayInternalIdentity, PlaneCaller
from veoveo_mcp.internal_auth import BEARER_SCOPE_KEY, IDENTITY_SCOPE_KEY
from veoveo_mcp.tasks import PrincipalKind, TaskOwner

from .app_state import AppState


def _invalid(message: str) -> McpError:
    return McpError(ErrorData(code=INVALID_REQUEST, message=message))


def identity_from_scope(scope: dict[str, Any]) -> GatewayInternalIdentity:
    identity = scope.get(IDENTITY_SCOPE_KEY)
    if identity is None:
        raise _invalid("gateway identity missing")
    return identity


def caller_from_scope(scope: dict[str, Any]) -> PlaneCaller:
    identity = identity_from_scope(scope)
    bearer = scope.get(BEARER_SCOPE_KEY)
    if bearer is None:
        raise _invalid("forwarded bearer missing")
    return PlaneCaller.from_identity(identity, bearer)


def request_scope(server: Any) -> dict[str, Any]:
    """The ASGI scope of the HTTP request behind the current MCP call."""
    request = server.request_context.request
    if request is None:
        raise _invalid("authenticated HTTP context missing")
    return request.scope


def runtime_owner(identity: GatewayInternalIdentity) -> TaskOwner:
    actor = identity.actor
    return TaskOwner(
        principal_key=actor.id,
        principal_kind=PrincipalKind(actor.kind.value),
        issuer=actor.issuer,
        subject=actor.subject,
        profile=identity.profile,
        tenant_key=actor.tenant,
        authority=identity.authority,
        data_labels=frozenset(actor.data_labels),
    )


def task_owner_allows(owner: TaskOwner, identity: GatewayInternalIdentity) -> bool:
    caller = runtime_owner(identity)
    return owner.allows(
        caller.principal_key,
        caller.profile,
        caller.tenant_key,
        caller.data_labels,
    )


async def require_task_owner(
    state: AppState, identity: GatewayInternalIdentity, task_id: str
) -> None:
    owner = await state.tasks.owner(task_id)
    if owner is None:
        raise _invalid("task ownership record missing")
    if not task_owner_allows(owner, identity):
        raise _invalid("datasheet task policy denied request")
